### PR TITLE
Updated oop_macro.nim with some improvements.

### PR DESCRIPTION
**Pull request by jyapayne**
Added out of order method execution inside the class.
Added super-like functionality to call parent methods from child methods.

### CODE REVIEW

File: content/oop_macro.md [hunk 1]
```diff
@@ -153,19 +208,18 @@ macro class*(head: expr, body: stmt): stmt {.immediate.} =
...
-      quote do:
-        type `typeName` = ref object
-    else:
-      quote do:
-        type `typeName` = ref object of `baseName`
-  )
+  var type_decl: PNimrodNode
+  if baseName == nil:
+    type_decl = quote do:
+      type `typeName` = ref object of RootObj
```
Comment (flaviut) [inline content/oop_macro.md]: oops, I forgot about this. Thanks.

File: content/oop_macro.md [hunk 2]
```diff
@@ -175,28 +229,79 @@ macro class*(head: expr, body: stmt): stmt {.immediate.} =
   #           ObjectTy
   #             Empty
   #             OfInherit
-  #               Ident !"RootObj"
+  #               Ident !"TObject"
```
Comment (flaviut) [inline content/oop_macro.md]: Seems like this an artifact from the merge.

File: content/oop_macro.md [hunk 3]
```diff
@@ -175,28 +229,79 @@ macro class*(head: expr, body: stmt): stmt {.immediate.} =
...
+  #  type 
   #    Animal = ref object of RootObj
   #      name: string
   #      age: int
   #
-  #  method vocalize(this: Animal): string =
+  #  method vocalize(this: Animal): string
+  #  method vocalizeAnimal(this: Animal): string   <= this is the magic for calling base class methods
+  #  method age_human_yrs(this: Animal): int
+  #  method age_human_yrsAnimal(this: Animal): int <= this is the magic for calling base class methods
```
Comment (flaviut) [inline content/oop_macro.md]: This isn't useful, [`procCall`](http://goran.krampe.se/2014/11/30/nim-and-oo-part-iv/) exists. It's quite new, and I'm not aware of it being documented anywhere though.

Comment (jyapayne) [inline content/oop_macro.md]: Ah, yeah, it took me a while to come up with this solution and I was looking for something like procCall. I can update the solution to work with it.

File: content/oop_macro.md (hunk 2 above)
Comment (jyapayne) [inline content/oop_macro.md]: Ah yep, my bad :)

**Comment by flaviut** [comment]
Sorry for taking so long to review the PR.

My goal here is to show off macros, not to create something usable. I don't feel like the changes help further explain the macro system, they simply add utility to it.

You should feel free to release the improved macro however you'd like though.

**Comment by jyapayne** [comment]
Ok, no worries.

I just thought it might be appropriate in the nim by example page. Thanks for the pointer about procCall :)

Cheers,
Joey

**Pull request closed without merging**